lib/Conversion/ImportAIGER/ImportAIGER.cpp

```diff
@@ -0,0 +1,982 @@
+    return emitError(lexer.getCurrentLoc(), message);
+  }
+
+  /// Expect and consume a specific token kind
+  ParseResult expectToken(AIGERTokenKind kind, const Twine &message);
+
+  /// Parse a number token into result
+  ParseResult parseNumber(unsigned &result);
+
+  /// Parse a binary encoded number (variable-length encoding)
+  ParseResult parseBinaryNumber(unsigned &result);
+
+  /// Expect and consume a newline token
+  ParseResult parseNewLine();
+};
+
+} // anonymous namespace
+
+//===----------------------------------------------------------------------===//
+// AIGERLexer Implementation
+//===----------------------------------------------------------------------===//
+
+void AIGERLexer::skipWhitespace() {
+  while (curPtr < curBuffer.end()) {
+    // NOTE: Don't use llvm::isSpace here because it also skips '\n.
+    if (*curPtr == ' ' || *curPtr == '\t' || *curPtr == '\r') {
+      ++curPtr;
+      continue;
+    }
+
+    // Treat "//" as whitespace. This is not part of the AIGER format, but we
+    // support it for FileCheck tests.
+    if (*curPtr == '/' &&
+        (curPtr + 1 < curBuffer.end() && *(curPtr + 1) == '/')) {
+      skipUntilNewline();
+      continue;
+    }
+    break;
+  }
+}
+
+void AIGERLexer::skipUntilNewline() {
+  while (curPtr < curBuffer.end() && *curPtr != '\n')
+    ++curPtr;
+  if (curPtr < curBuffer.end() && *curPtr == '\n')
+    ++curPtr;
+}
+
+AIGERToken AIGERLexer::lexNumber() {
+  const char *start = curPtr;
+  while (curPtr < curBuffer.end() && llvm::isDigit(*curPtr))
+    ++curPtr;
+  return makeToken(AIGERTokenKind::Number, start);
+}
+
+AIGERToken AIGERLexer::lexIdentifier() {
+  const char *start = curPtr;
+  while (curPtr < curBuffer.end() && (llvm::isAlnum(*curPtr) || *curPtr == '_'))
+    ++curPtr;
+
+  StringRef spelling(start, curPtr - start);
+  AIGERTokenKind kind = AIGERTokenKind::Identifier;
+
+  return makeToken(kind, start);
+}
+
+AIGERToken AIGERLexer::nextToken() {
+  skipWhitespace();
+
+  auto impl = [this]() {
+    if (curPtr >= curBuffer.end())
+      return makeToken(AIGERTokenKind::EndOfFile, curPtr);
+
+    const char *start = curPtr;
+    char c = *curPtr++;
+
+    switch (c) {
+    case '\n':
+      return makeToken(AIGERTokenKind::Newline, start);
+    case '\r':
+    case ' ':
+    case '\t':
+      llvm_unreachable("Whitespace should have been skipped");
+      return makeToken(AIGERTokenKind::Error, start);
+    default:
+      if (llvm::isDigit(c)) {
+        --curPtr; // Back up to re-lex the number
+        return lexNumber();
+      }
+      if (llvm::isAlpha(c) || c == '_') {
+        --curPtr; // Back up to re-lex the identifier
+        return lexIdentifier();
+      }
+      // Not officially part of the AIGER format, but we support it for
+      // FileCheck tests.
+      if (c == '/') {
+        if (*curPtr == '/') {
+          llvm_unreachable("// should have been skipped");
+          return makeToken(AIGERTokenKind::Error, start);
+        }
+      }
```

Comment:
Could you just assume that `//` has already been skipped? I'm wondering if `llvm_unreachable` is worth the additional check.